elifepubmed/generate.py

```diff
@@ -601,6 +601,45 @@ def set_grants(parent, poa_article):
                 set_object(parent, "grant", params)
 
 
+def dataset_assigning_authority(assigning_authority, uri):
+    """precise assigning_authority value considering the uri in some cases"""
+    authority_map = OrderedDict([('NCBI', [])])
+    # tuple in form: uri hint, new assigning authority
+    authority_map['NCBI'].append(('www.ncbi.nlm.nih.gov/geo', 'NCBI:geo'))
+    authority_map['NCBI'].append(('www.ncbi.nlm.nih.gov/projects/gap', 'NCBI:dbgap'))
+    authority_map['NCBI'].append(('www.ncbi.nlm.nih.gov/nuccore', 'NCBI:nucleotide'))
+    authority_map['NCBI'].append(('www.ncbi.nlm.nih.gov/sra', 'NCBI:sra'))
+
+    if assigning_authority in authority_map:
+        for hint, new_value in authority_map.get(assigning_authority):
+            if hint in uri:
+                assigning_authority = new_value
+    return assigning_authority
```

Comment:
I'm not sure what `assigning_authority` is but could the dictionary have keys that you could perform O(1) looks ups on?

Reply:
If I am considering the question correctly, I don't think think there's a more efficient way, except to break the loop after finding a value. Maybe this function is not very clear what it does? 

The optimisation I added was if the `assigning_authority` is not `NCBI`, then we skip the iteration entirely. If it is `NCBI`, then we look for a string fragment in the `uri` to determine the more precise assigning_authority value, based on the uri.

Please be welcome to try it in a different way that satisfies the existing test scenarios. It is an interesting problem to solve.

Reply:
Sorry @gnott, I'm not following. Would you mind giving me an example of the value of `assigning_authority` and `uri`?

Reply:
I believe the answer is in the tests, example uri `https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE48760`.

The way I read it, it is determining the `assigning_authority`, such as `NCBI:geo` from the `uri`. If none of the `uri hints` match it is returning the passed in `assigning_authority` (which maybe could be moved to the end and called `default_assigning_authority` or similar). If `https://` was added to the `uri hint` it could use `startswith` but would assume it's `https` or it could parse the `uri`. Not sure how often this would be called though.

Reply:
Hey @de-code, I got a little lost on the suggestion about `default_assigning_authority`, but I changed it so the new value is returned immediately when found, which might make naming a default not required (I'm not completely sure).